common/factorization/par_ilu_kernels.hpp.inc

```diff
@@ -33,6 +33,203 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 namespace kernel {
 
 
+namespace detail {
+
+
+// Default implementation for the unsorted case
+template <bool IsSorted>
+struct find_helper {
+    template <typename Group, typename IndexType>
+    static __device__ bool find(Group subwarp_grp, const IndexType *first,
+                                const IndexType *last, const IndexType &value)
+    {
+        auto subwarp_idx = subwarp_grp.thread_rank();
+        bool found{false};
+        for (auto curr_start = first; curr_start < last;
+             curr_start += subwarp_grp.size()) {
+            const auto curr = curr_start + subwarp_idx;
+            found = (curr < last && *curr == value);
+            found = subwarp_grp.any(found);
+            if (found) {
+                break;
+            }
+        }
+        return found;
+    }
+};
+
+
+// Improved version in case the CSR matrix is sorted
+template <>
+struct find_helper<true> {
+    template <typename Group, typename IndexType>
+    static __device__ bool find(Group subwarp_grp, const IndexType *first,
+                                const IndexType *last, const IndexType &value)
+    {
+        const auto length = static_cast<IndexType>(last - first);
+        const auto pos =
+            group_wide_search(IndexType{}, length, subwarp_grp,
+                              [&](IndexType i) { return first[i] >= value; });
+        return pos < length && first[pos] == value;
+    }
+};
+
+
+}  // namespace detail
+
+
+// SubwarpSize needs to be a power of 2
+// Each subwarp works on one row
+template <bool IsSorted, int SubwarpSize, typename IndexType>
+__global__
+    __launch_bounds__(default_block_size) void find_missing_diagonal_elements(
+        IndexType num_rows, IndexType num_cols,
+        const IndexType *__restrict__ col_idxs,
+        const IndexType *__restrict__ row_ptrs,
+        IndexType *__restrict__ elements_to_add_per_row,
+        bool *__restrict__ changes_required)
+{
+    const auto total_thread_count =
+        static_cast<size_type>(blockDim.x) * gridDim.x / SubwarpSize;
+    const auto tidx =
+        threadIdx.x + static_cast<size_type>(blockIdx.x) * blockDim.x;
+    const auto begin_row = static_cast<IndexType>(tidx / SubwarpSize);
+
+    auto thread_block = group::this_thread_block();
+    auto subwarp_grp = group::tiled_partition<SubwarpSize>(thread_block);
+    const auto subwarp_idx = subwarp_grp.thread_rank();
+
+    bool local_change{false};
+    for (IndexType row = begin_row; row < num_rows; row += total_thread_count) {
+        if (row >= num_cols) {
```

Comment:
I think it would be nice to have. But in any case, if this works only for square matrices, we should probably add a check before calling this kernel ?

Reply:
In the `generate_l_u` ([here](https://github.com/ginkgo-project/ginkgo/blob/develop/core/factorization/par_ilu.cpp#L77)), we do check if it is square or not.
Here, I simply thought it is extremely easy to also support non-square matrices, so I did it.